### PR TITLE
refactor(json-v6): implement waitForResume and waitForDuration in engine

### LINKED ISSUE #9472: feat(engine): consume and support `delay`, `waitForDuration`, and `waitForResume` commands from JSONv6
## Overview

As a JSONv6 protocol user author, I want to instruct the protocol to wait...

- for a period of time, or
- until I click resume

...before proceeding.

In the JSONv6 schema, these behaviors are accomplished by the following commands:

- `delay`
- `waitForResume` (see #10835)
- `waitForDuration` (see #10835)

## Acceptance criteria

- [ ] ProtocolEngine has `waitForResume` command that matches JSONv6 schema
    - Implementation pauses execution until a `resume` action is received
- [ ] ProtocolEngine has a `waitForTime` command that matches JSONv6 schema
    - Implementation pauses execution the given time elapses
- [ ] Both `waitForResume` and `waitForTime` are overridden by a `stop` action 
- [ ] `pause` / `resume` actions from the client do not affect the time delay
    - In other words, a `waitForTime` always waits at least the given time
    - If a pause action is issued while a time delay is in progress, the protocol should remain paused once the time elapses 
- [ ] JSONv6 > ProtocolEngine translation layer translates `delay` commands into `waitForResume` and `waitForTime`

## Implementation notes

- `waitForResume` can be implemented as a rename of the ProtocolEngine's existing `pause` command
- The `waitForDuration` spec above _seems_ complicated, but it's not! Details below

### `waitForDuration` implementation notes

The goal with `waitForDuration` is to preserve the existing v5.x.x behavior. Some examples of that behavior - given a `waitForDuration` of 30 seconds:

- If, while the delay is ongoing, a user sends a pause, the delay’s internal timer still proceeds. Once the delay’s 30 seconds have passed, the protocol stays paused by the user-initiated pause; sending a resume will then resume the protocol.
- If, while the delay is in progress you send a pause 10 seconds in, wait for 5 seconds and then send a resume, the protocol will delay for another 15 seconds and then proceed to the next step. So the 30 second delay always completes.

This may seem daunting, but the ProtocolEngine's existing architecture is already geared exactly towards this! **If we implement `waitForDuration` as a simple call to `await anyio.sleep()`** (or asyncio, whatever), I believe we'll get the spec'd behavior "for free".

If a `waitForDuration` is ongoing in a call to `await sleep(...)`:

- If a `engine.pause()` is called, the status will change to `paused` and the queue worker will be prevented from picking up **new** commands, but the currently running command will not be affected
     - In other words, the `sleep` timer will keep ticking
    - When the `sleep` timer elapses, the engine will not proceed to the next command, because the queue worker is still paused from the `engine.pause()`
    - A call to `engine.resume()` will unblock the queue worker, allowing the protocol to proceed
- If a call to `engine.resume()` happens while the `sleep` timer is still ticking, the queue worker will _also_ not move on, because it will still be waiting for the `waitForTime` command to complete

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR adds the `waitForDuration` command to the engine and replaces the `pause` command with the the `waitForResume` command. Some changes were necessary across the stack to maintain backwards compatibility and support display of protocol pauses when using JSONv6 protocols.

Closes #9472. See ticket for acceptance criteria and more detailed command behavior specs

## Changelog

- Rename `pause` command to `waitForResume` in ProtocolEngine
    - Continue to accept `pause` as the command type for backwards compatibility with both the front-end **and persisted protocol analyses and runs**
    - Add `waitForResume` to various places in the app where `commandType: pause` was checked to see if a command paused the protocol
    - **Note**: I did not do the same for `waitForDuration`, because the app does not currently look for `delay` commands, nor does the `LegacyCommandMapper` map them 
- Merge `pause` command and `waitForResume` command in JSONv6 schema

## Review requests

- Does back compat support make sense?
- Do these changes make any sense to _also_ land in `release_6.0.0`? 

### Smoke test

A cool thing about this PR is that it doesn't require actual robot hardware to validate waits! Thanks to the `ProtocolEngine`, hardware simulation is completely separate from protocol analysis, so during actual runs of JSONv6 protocols, pauses and delays will behave as normal in the dev server (or other `ENABLE_VIRTUAL_SMOOTHIE` environments).

Obviously, though, if you _can_ test on an actual OT-2, that's always better.

- [ ] Running a PAPIv2 protocol with pauses continues to function normally w.r.t. the run log
- [ ] A JSONv6 protocol (generated from PD edge or from fixtures) is able to pause and delay
- [ ] Previously persisted runs and analyses that used `pause` commands are still able to be returned by the HTTP API
    1. Checkout (dev server) or push (real hardware) `edge`
    2. Run a simple PAPIv2 protocol with a `protocol.pause()`
    3. Check that `GET /runs/:run_id/commands` contains a `pause` command
    4. Checkout or push `pe_pause-and-delay`
    5. Restart the server
    6. Check that `GET /runs/:run_id/commands` _still_ contains a `pause` command

```json
{
  "$otSharedSchema": "#/protocol/schemas/6",
  "schemaVersion": 6,
  "metadata": {
    "protocolName": "Simple test protocol",
    "author": "engineering <engineering@opentrons.com>",
    "description": "A short test protocol",
    "created": 1223131231,
    "tags": ["unitTest"]
  },
  "robot": {
    "model": "OT-2 Standard",
    "deckId": "ot2_standard"
  },
  "pipettes": {},
  "modules": {},
  "labware": {},
  "liquids": {},
  "labwareDefinitions": {},
  "commands": [
    {
      "commandType": "waitForResume",
      "params": {
        "message": "Waiting until you click resume"
      }
    },
    {
      "commandType": "waitForDuration",
      "params": {
        "message": "Waiting for 30 seconds",
        "seconds": 30
      }
    }
  ],
  "commandAnnotations": []
}
```

## Risk assessment

Medium. This touches on some fiddly backwards compatibility stuff

### CODE REVIEW

File: app/src/organisms/RunDetails/hooks.ts [hunk 1]
```diff
@@ -58,7 +59,7 @@ export function useTimeElapsedSincePause(): string | null {
     runRecord?.data &&
     ['paused', 'pause-requested'].includes(runRecord?.data.status) &&
     mostRecentAction != null &&
-    mostRecentAction?.actionType === 'pause'
+    ['pause', 'waitForResume'].includes(mostRecentAction?.actionType)
```
Comment (b-cooper) [inline app/src/organisms/RunDetails/hooks.ts]: just to make sure, this is checking for an `action` with `actionType === 'waitForResume'` is "waitForResume" both a new `actionType` and a new `commandType`?

Comment (mcous) [reply app/src/organisms/RunDetails/hooks.ts]: Nope, misread the code. Will fix

**Comment by mcous** [comment]
So far, I smoke tested the following:

**On an OT-2**

- A JSONv6 protocol with `delay` commands 
    - [x] Protocol analyzes 
    - [x] `delay` is translated to `waitForResume` and `waitForDuration`
    - [x] Correctly waits for resume
    - [x] Correctly waits for the requested duration
    - [x] Is able to be cancelled
- A JSONv6 protocol with `waitForResume` and `waitForDuration` commands
    - [x] Protocol analyzes 
    - [x] Correctly waits for resume
    - [x] Correctly waits for the requested duration
    - [x] Is able to be cancelled

**On a dev server**

- [x] Persisted protocols with `pause` are still able to be retrieved on this branch
- [x] `waitForDuration` adhered to spec in #9472; user initiated `play`/`pause` actions did not interfere with the timing of the `waitForDuration`

File: api/src/opentrons/protocol_api_experimental/protocol_context.py [hunk 2]
```diff
@@ -226,7 +226,7 @@ def pause(self, msg: Optional[str] = None) -> None:
             msg: An optional message to attach to the pause command. This message
                 will be available to the app or client controlling the protocol.
         """
```
Comment (X-sam) [inline api/src/opentrons/protocol_api_experimental/protocol_context.py:228]: should we put a deprecation notice in this docstring?

**Review by X-sam — APPROVED** [review]
lgtm

File: api/src/opentrons/protocol_engine/commands/wait_for_duration.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+class WaitForDurationImplementation(
+    AbstractCommandImpl[WaitForDurationParams, WaitForDurationResult]
+):
+    """Wait for duration command implementation."""
+
+    def __init__(self, run_control: RunControlHandler, **kwargs: object) -> None:
+        self._run_control = run_control
+
+    async def execute(self, params: WaitForDurationParams) -> WaitForDurationResult:
+        """Wait for a duration of time using."""
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/wait_for_duration.py]: Incomplete sentence?

File: api/src/opentrons/protocol_engine/state/commands.py [hunk 4]
```diff
@@ -183,6 +183,7 @@ def handle_action(self, action: Action) -> None:  # noqa: C901
                 id=action.command_id,
                 key=action.command_key,
                 createdAt=action.created_at,
+                commandType=action.request.commandType,  # type: ignore[arg-type]
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/commands.py]: Can we leave a note here that even though `_CommandCls.commandType` probably has a default, it's not necessarily the exact string given in the `queued_command`?

Actually, what happens if we omit this?

**Review by b-cooper — APPROVED** [review]
⏲️ ⏹️

File: api/src/opentrons/protocol_api_experimental/protocol_context.py (hunk 2 above)
Comment (mcous) [reply api/src/opentrons/protocol_api_experimental/protocol_context.py:228]: Good question! Short version: for this "public" API, I think "not yet".

Long version: both PD and PAPIv2 use the word "pause" in their UI and public interface, respectively. For now I think we shouldn't touch the proposed PAPIv3 method, but there's a strong argument to adding the rename and depreciation notice to PAPIv2 so we can rename the method in PAPIv3

**Review by SyntaxColoring — APPROVED** [review]
Python code changes look good to me.

File: api/src/opentrons/protocol_engine/state/commands.py (hunk 4 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/state/commands.py]: If we omit this, then a `WaitForResumeCreate` with `pause` will be mapped to a `WaitForResume` with `commandType="waitForResume"` which is... actually not the worst idea now that I think about it. Will remove this I think, but keep `pause` as a valid `commandType` for deserialization-from-DB purposes and update the mapping test accordingly

**Comment by mcous** [comment]
Confirmed that latest commit did not affect command deserialization from persistence layer, will merge once CI is green

**Pull request merged**